converter.go

```diff
@@ -521,12 +652,153 @@ func arrowToValue(rowIdx int, srcColumnMeta execResponseRowType, srcValue arrow.
+					decs := make([]float64, offsets[rowIdx+1]-offsets[rowIdx])
+					for j := offsets[rowIdx]; j < offsets[rowIdx+1]; j++ {
+						v := arrowDecimal128ToValue(values.(*array.Decimal128), int(j), higherPrecision, srcColumnMeta.Fields[0].Scale)
+						if decs[j-offsets[rowIdx]], err = strconv.ParseFloat(v.(string), 64); err != nil {
+							return nil, err
+						}
+					}
+					return decs, nil
+				}
+			case *array.Int64:
+				ints := make([]int64, offsets[rowIdx+1]-offsets[rowIdx])
+				for j := offsets[rowIdx]; j < offsets[rowIdx+1]; j++ {
+					ints[j-offsets[rowIdx]] = arrowInt64ToValue(values.(*array.Int64), int(j), higherPrecision, srcColumnMeta.Fields[0].Scale).(int64)
+				}
+				return ints, nil
+			case *array.Int32:
+				ints := make([]int32, offsets[rowIdx+1]-offsets[rowIdx])
+				for j := offsets[rowIdx]; j < offsets[rowIdx+1]; j++ {
+					ints[j-offsets[rowIdx]] = arrowInt32ToValue(values.(*array.Int32), int(j), higherPrecision, srcColumnMeta.Fields[0].Scale).(int32)
+				}
+				return ints, nil
+			case *array.Int16:
+				ints := make([]int16, offsets[rowIdx+1]-offsets[rowIdx])
+				for j := offsets[rowIdx]; j < offsets[rowIdx+1]; j++ {
+					ints[j-offsets[rowIdx]] = arrowInt16ToValue(values.(*array.Int16), int(j), higherPrecision, srcColumnMeta.Fields[0].Scale).(int16)
+				}
+				return ints, nil
+			case *array.Int8:
+				ints := make([]int8, offsets[rowIdx+1]-offsets[rowIdx])
+				for j := offsets[rowIdx]; j < offsets[rowIdx+1]; j++ {
+					ints[j-offsets[rowIdx]] = arrowInt8ToValue(values.(*array.Int8), int(j), higherPrecision, srcColumnMeta.Fields[0].Scale).(int8)
+				}
+				return ints, nil
+			}
+		case realType:
+			floats := make([]float64, offsets[rowIdx+1]-offsets[rowIdx])
+			for j := offsets[rowIdx]; j < offsets[rowIdx+1]; j++ {
+				floats[j-offsets[rowIdx]] = arrowRealToValue(values.(*array.Float64), int(j)).(float64)
+			}
+			return floats, nil
+		case textType:
+			strings := make([]string, offsets[rowIdx+1]-offsets[rowIdx])
+			for j := offsets[rowIdx]; j < offsets[rowIdx+1]; j++ {
+				strings[j-offsets[rowIdx]] = values.(*array.String).Value(int(j))
+			}
+			return strings, nil
+		case booleanType:
+			bools := make([]bool, offsets[rowIdx+1]-offsets[rowIdx])
+			for j := offsets[rowIdx]; j < offsets[rowIdx+1]; j++ {
+				bools[j-offsets[rowIdx]] = values.(*array.Boolean).Value(int(j))
+			}
+			return bools, nil
+		case binaryType:
+			binaries := make([][]byte, offsets[rowIdx+1]-offsets[rowIdx])
+			for j := offsets[rowIdx]; j < offsets[rowIdx+1]; j++ {
+				binaries[j-offsets[rowIdx]] = values.(*array.Binary).Value(int(j))
+			}
+			return binaries, nil
+		case dateType:
+			dates := make([]time.Time, offsets[rowIdx+1]-offsets[rowIdx])
+			for j := offsets[rowIdx]; j < offsets[rowIdx+1]; j++ {
+				dates[j-offsets[rowIdx]] = arrowDateToValue(values.(*array.Date32), int(j)).(time.Time)
+			}
+			return dates, nil
+		case timeType:
+			times := make([]time.Time, offsets[rowIdx+1]-offsets[rowIdx])
```

Comment:
there is a lot of potentially generic logic (iterating through data and offsets). Can we somehow extract this to avoid writing so many loops?

Reply:
I refactored it a bit. Not sure if it is better though.